NEON 41 - Added support for scroll search API

jacklechnercaci opened the pull request.

jacklechnercaci commented:
I see now that the issue on jira has "if the limit is > 10,000". I need to add the case that this is used if the limit is above that threshold. Right now it always tries this... :-X

ThomasSchellenbergNextCentury commented on esadapter/src/main/java/com/ncc/neon/adapters/es/ElasticsearchResultsConverter.java:
```diff
@@ -94,7 +100,26 @@ public static TabularQueryResult convertResults(Query query, SearchResponse resp
+        List<Map<String, Object>> allHits = new ArrayList<>();
+        allHits.addAll(ElasticsearchResultsConverter.extractHitsFromResults(response));
+
+        while (response.getHits().getHits() != null && response.getHits().getHits().length > 0) {
```
@jacklechnercaci Beside what you already said about only using the search scroll API if the limit is greater than 10K, we should also only return a number of results equal to or less than the limit.  For example, if my limit is 15K, and I have 20K records in my datastore, currently this method returns all 20K records, and not just the first 15K.

jacklechnercaci replied:
- Updated the PR so that if the limit is over 10k, it does a scrolled request.
- Now returns the number of records equal to the limit.

ThomasSchellenbergNextCentury commented on esadapter/src/main/java/com/ncc/neon/adapters/es/ElasticsearchResultsConverter.java:
```diff
@@ -94,7 +100,26 @@ public static TabularQueryResult convertResults(Query query, SearchResponse resp
+        allHits.addAll(ElasticsearchResultsConverter.extractHitsFromResults(response));
+
+        while (response.getHits().getHits() != null && response.getHits().getHits().length > 0) {
+            SearchScrollRequest scrollRequest = new SearchScrollRequest(scrollId);
```
@jacklechnercaci Do we know how many hits each scroll request returns?  The [documentation](https://www.elastic.co/guide/en/elasticsearch/client/java-rest/6.8/java-rest-high-search-scroll.html#_initialize_the_search_scroll_context) says that the request size is set with the `SearchSourceBuilder`, which we're setting to the query limit in the [ElasticsearchQueryConverter](https://github.com/NextCenturyCorporation/nucleus-data-server/blob/master/esadapter/src/main/java/com/ncc/neon/adapters/es/ElasticsearchQueryConverter.java#L333-L334) (and I'm guessing that ignores the normal 10K maximum?).  So I believe if we query for 15K results, it looks like this just runs a single scroll request for all 15K.  If this is correct, I'd think it may be better for performance (and also less likely to crash) if we run multiple smaller queries (maybe 1K each).  Unless I'm misunderstanding something?

jacklechnercaci replied:
There's a config `index.max_result_window` that is defaulted to 10k for performance reasons and it's suggested not to be modified. This means that the max batch size is 10k, but I found `terminateAfter()` in the SearchSourceBuilder that does 10k batches until it reaches the size set there. This should be performant and it does what we wanted.

ThomasSchellenbergNextCentury replied:
Ah, OK.  So if I want 15K, it will run a request for 10K and then a request for 5K?  That all seems fine.  I'd be interested to know how one 10K request compares to, say, 10 sequential 1K requests, but it would make sense for the 10K to be faster (less overhead).

jacklechnercaci replied:
Yep, that's exactly how it works. From what I've seen online, everyone prefers 10k batches.

ThomasSchellenbergNextCentury commented:
@jacklechnercaci Some tests are failing

jacklechnercaci commented:
Which tests were failing? I just ran all tests and got 46 of 46 passes.

ThomasSchellenbergNextCentury commented:
@jacklechnercaci You may need to do a clean build: `./gradlew clean build`

```
> Task :esadapter:test
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended

com.ncc.neon.adapters.es.ElasticsearchQueryConverterTest > convertQueryLimitMaximumTest FAILED
    org.opentest4j.AssertionFailedError at ElasticsearchQueryConverterTest.java:946

com.ncc.neon.adapters.es.ElasticsearchQueryConverterTest > convertQueryLimitAndOffsetTest FAILED
    org.opentest4j.AssertionFailedError at ElasticsearchQueryConverterTest.java:805

com.ncc.neon.adapters.es.ElasticsearchQueryConverterTest > convertQueryAdvancedTest FAILED
    org.opentest4j.AssertionFailedError at ElasticsearchQueryConverterTest.java:820

com.ncc.neon.adapters.es.ElasticsearchQueryConverterTest > convertQueryLimitTest FAILED
    org.opentest4j.AssertionFailedError at ElasticsearchQueryConverterTest.java:785
```

jacklechnercaci commented:
The tests were using a `SearchSourceBuilder` that didn't have `terminateAfter()` set. I corrected this. I also changed the query converter to determine the terminateAfter value separate from the `getLimit()` method because the `supportsUnlimited` phrase has meaning to aggregate searches.

ThomasSchellenbergNextCentury commented:
This is super weird... It looks like I'm still getting all 32K results back even when I set the limit to 15K... I verified that I have your latest code, so I'm not sure what's wrong... Here's an example query (from the data table, with limit = 15K):

```
{
  "whereClause": {
    "type": "where",
    "lhs": {
      "database": "earthquakes",
      "table": "_doc",
      "field": "time"
    },
    "operator": "!=",
    "rhs": null
  },
  "aggregateClauses": [],
  "groupByClauses": [],
  "orderByClauses": [
    {
      "type": "field",
      "fieldClause": {
        "database": "earthquakes",
        "table": "_doc",
        "field": "time"
      },
      "order": -1
    }
  ],
  "limitClause": {
    "limit": 15000
  },
  "offsetClause": {
    "offset": 0
  },
  "joinClauses": [],
  "isDistinct": false,
  "selectClause": {
    "database": "earthquakes",
    "table": "_doc",
    "fieldClauses": [
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "time"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "depth"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "_id"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "id"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "latitude"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "longitude"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "mag"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "magType"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "net"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "time"
      },
      {
        "database": "earthquakes",
        "table": "_doc",
        "field": "type"
      }
    ]
  }
}
```

jacklechnercaci commented:
I tried that exact query and got 15k results back.
